runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowExecutionContext.java

```diff
@@ -313,19 +325,24 @@ public Closeable enterState(ExecutionState newState) {
       if (isDataflowProcessElementState) {
         DataflowExecutionState newDFState = (DataflowExecutionState) newState;
         if (newDFState.getStepName() != null && newDFState.getStepName().userName() != null) {
-          if (this.activeMessageMetadata != null) {
-            recordActiveMessageInProcessingTimesMap();
+          synchronized (this) {
+            if (this.activeMessageMetadata != null) {
+              recordActiveMessageInProcessingTimesMap();
+            }
+            this.activeMessageMetadata =
+                ActiveMessageMetadata.create(
+                    newDFState.getStepName().userName(), clock.getMillis());
           }
-          this.activeMessageMetadata =
-              ActiveMessageMetadata.create(newDFState.getStepName().userName(), clock.getMillis());
         }
         elementExecutionTracker.enter(newDFState.getStepName());
       }
 
       return () -> {
         if (isDataflowProcessElementState) {
-          if (this.activeMessageMetadata != null) {
-            recordActiveMessageInProcessingTimesMap();
+          synchronized (this) {
+            if (this.activeMessageMetadata != null) {
```

Comment:
```suggestion
```

check can be removed since recordActiveMessageInProcessingTimesMap has a null check inside.
Removing the check also allows removing the `synchronized (this) {` since recordActiveMessageInProcessingTimesMap synchronizes.